azurerm_monitor_action_group - Add optional location parameter and default to Global #20151

mikemadeja opened the pull request:
- Add location parameter to this issue: https://github.com/hashicorp/terraform-provider-azurerm/issues/20151 #20151
  - Location defaults to **global** to avoid breaking current configs for users
  - Location forces new one to be created when changing locations
  - Updated document to reflect the new parameter

mikemadeja commented:
Hello @stephybun, may I ask what is breaking in regards to this change? The original code defaults to "Global", which I have defaulted to. But now you can add location and set it any of the regions the API offers.

stephybun commented on internal/services/monitor/monitor_action_group_resource.go:
```diff
@@ -58,6 +58,13 @@ func resourceMonitorActionGroup() *pluginsdk.Resource {
 
 			"resource_group_name": commonschema.ResourceGroupName(),
 
+			"location": {
+				Type:     pluginsdk.TypeString,
+				Optional: true,
+				Default:  "Global",
+				ForceNew: true,
+			},
```
We have a commonschema function for `location` which we would typically use if this weren't a unique case, we should however still use parts of it. `Global` also needs to be lowercased since that's the normalisation location undergoes in the provider.

```suggestion
					"location": {
						Type:             schema.TypeString,
						Optional:         true,
						ForceNew:         true,
						Default:          "global",
						ValidateFunc:     validation.Any(
							location.EnhancedValidate,
							validation.StringInSlice([]string{
								"global",
							}, false),
						),
						StateFunc:        location.StateFunc,
						DiffSuppressFunc: location.DiffSuppressFunc,
					},
```

mikemadeja replied:
Thanks for the reply @stephybun, I have added what you have wrote. I needed to add the schema dependency. 
	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"

stephybun commented on internal/services/monitor/monitor_action_group_resource.go:
```diff
@@ -476,6 +483,7 @@ func resourceMonitorActionGroupCreateUpdate(d *pluginsdk.ResourceData, meta inte
 	client := meta.(*clients.Client).Monitor.ActionGroupsClient
 	tenantId := meta.(*clients.Client).Account.TenantId
 	subscriptionId := meta.(*clients.Client).Account.SubscriptionId
+	location := d.Get("location").(string)
```
We need to normalise this
```suggestion
	location := location.Normalize(d.Get("location").(string))
```

mikemadeja replied:
Thanks @stephybun, this has been updated as well. Still learning...

stephybun reviewed (changes requested):
Thanks for this PR @mikemadeja!

I pre-emptively marked this as a breaking change under the assumption that we would use the same schema for location that we have elsewhere in the provider, however this turns out to be a unique case since it has defaulted to `global` so far.

In addition to the comments left in-line we would also need to address the following points before we can get this merged:
1. Normalise and set the location back to state in the read
2. Add a test where location is set

Once that's done we can take another look through and hopefully get this merged.

mikemadeja commented:
Hello @stephybun, I added a test with location. But I am researching if anything else needs to be done to utilize those tests. 

Under internal/services/monitor/monitor_action_group_resource_test.go, I created a function called locationBasic and added a location field and set it to global. The test then expects global.

katbyte reviewed (changes requested):
all tests are failing with 

```------- Stdout: -------
=== RUN   TestAccMonitorActionGroup_emailReceiver
=== PAUSE TestAccMonitorActionGroup_emailReceiver
=== CONT  TestAccMonitorActionGroup_emailReceiver
    testcase.go:110: Step 2/2 error running import: ImportStateVerify attributes not equivalent. Difference is shown below. The - symbol indicates attributes missing after import.
        
          map[string]string{
        - 	"location": "global",
          }
--- FAIL: TestAccMonitorActionGroup_emailReceiver (191.17s)
FAIL
```

 i see that you pushed a fix 7 hours ago, that maybe have been just before/after i triggers them! the import state verify failing usually means there is something off in your read function, not saving location to state.

mikemadeja commented:
Thanks @katbyte and @stephybun. I made an error on one of the tests I copied and have updated. I also compiled it locally and verified the state is there so hopefully it was my error on the tests. May I ask where I can see the tests? Would it be this one? https://github.com/hashicorp/terraform-provider-azurerm/actions/runs/4263642693/jobs/7420650522

![image](https://user-images.githubusercontent.com/28938328/221213494-cf9cc63e-f2bc-4618-9a6a-8f1f281816a5.png)

mikemadeja commented:
Hello @katbyte and @stephybun. Just wanted to see if there were any other concerns with this PR? Just want to make sure I didn't miss anything. :-)

Thanks,

Mike

stephybun commented on internal/services/monitor/monitor_action_group_resource_test.go:
```diff
@@ -300,6 +300,38 @@ func TestAccMonitorActionGroup_disabledUpdate(t *testing.T) {
 	})
 }
 
+func TestAccMonitorActionGroup_locationBasic(t *testing.T) {
+	data := acceptance.BuildTestData(t, "azurerm_monitor_action_group", "test")
+	r := MonitorActionGroupResource{}
+
+	data.ResourceTest(t, r, []acceptance.TestStep{
+		{
+			Config: r.locationBasic(data),
+			Check: acceptance.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+				check.That(data.ResourceName).Key("location").HasValue("global"),
+			),
+		},
+		data.ImportStep(),
+		{
+			Config: r.basic(data),
+			Check: acceptance.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+				check.That(data.ResourceName).Key("location").HasValue("global"),
+			),
+		},
+		data.ImportStep(),
+		{
+			Config: r.locationBasic(data),
+			Check: acceptance.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+				check.That(data.ResourceName).Key("location").HasValue("global"),
+			),
+		},
+		data.ImportStep(),
+	})
+}
```
Since this property is `ForceNew` applying multiple configurations in the test will just end up recreating the resource at every single apply which isn't testing the new behaviour that's being added, that location can be set to a value other than `global`.

We should test that location can be set to something other than `global` since that's the default.

```suggestion
func TestAccMonitorActionGroup_location(t *testing.T) {
	data := acceptance.BuildTestData(t, "azurerm_monitor_action_group", "test")
	r := MonitorActionGroupResource{}

	data.ResourceTest(t, r, []acceptance.TestStep{
		{
			Config: r.location(data),
			Check: acceptance.ComposeTestCheckFunc(
				check.That(data.ResourceName).ExistsInAzure(r),
			),
		},
		data.ImportStep(),
	})
}
```

stephybun commented on internal/services/monitor/monitor_action_group_resource_test.go:
```diff
@@ -1163,3 +1195,24 @@ func (t MonitorActionGroupResource) Exists(ctx context.Context, clients *clients
 	return utils.Bool(resp.ID != nil), nil
 }
+
+func (MonitorActionGroupResource) locationBasic(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+provider "azurerm" {
+  features {}
+}
+
+resource "azurerm_resource_group" "test" {
+  name     = "acctestRG-%d"
+  location = "%s"
+}
+
+resource "azurerm_monitor_action_group" "test" {
+  name                = "acctestActionGroup-%d"
+  resource_group_name = azurerm_resource_group.test.name
+  short_name          = "acctestag"
+  enabled             = false
+  location	          = "global"
+}
+`, data.RandomInteger, data.Locations.Primary, data.RandomInteger)
+}
```
```suggestion
func (MonitorActionGroupResource) location(data acceptance.TestData) string {
	return fmt.Sprintf(`
provider "azurerm" {
  features {}
}

resource "azurerm_resource_group" "test" {
  name     = "acctestRG-%d"
  location = "%s"
}

resource "azurerm_monitor_action_group" "test" {
  name                = "acctestActionGroup-%d"
  resource_group_name = azurerm_resource_group.test.name
  short_name          = "acctestag"
  enabled             = false
  location	          = "swedencentral"
}
`, data.RandomInteger, data.Locations.Primary, data.RandomInteger)
}
```

stephybun reviewed (changes requested):
Hi @mikemadeja

The GHA link you're referring to runs the unit tests in the provider. When we post test failures we're referring to the acceptance tests which are run in our CI tool. You're able to run these tests as well, just be aware that they will incur costs on your subscription since they spin up resources in Azure.

In addition to the comments on the tests we still need to set the value of location into the state in the read function, we do that by adding the following line:

`d.Set("location", location.NormalizeNilable(resp.Location))`

Here:

https://github.com/hashicorp/terraform-provider-azurerm/blob/0186544d677c8dc94c00e74a15a2873447ee1b77/internal/services/monitor/monitor_action_group_resource.go#L590-L592

Once these points have been resolved this should be good to go, thanks!

mikemadeja commented:
Thanks @stephybun, I appreciate the help. Thought this was an easy change so I'm learning a lot on how all this works.

stephybun reviewed (approved):
Thanks @mikemadeja LGTM 🎉 

I hope you don't mind I pushed two small changes to your branch that put the property in the correct order in the docs which I overlooked in the review, as well as to format the Terraform config in the test. Thanks again for your work on this!

mikemadeja commented:
Thanks @stephybun, no problem.
